src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.X.cs

```diff
@@ -34,11 +33,21 @@ private static bool TryFormatUInt64X(ulong value, byte precision, bool useLower,
             // directly with nibbles. There may be opportunity for improvement by special-
             // casing output lengths of 2, 4, 8, and 16 and running them down optimized
             // code paths.
-
-            while ((uint)(--computedOutputLength) < (uint)destination.Length)
+            if (useLower)
+            {
+                while ((uint)(--computedOutputLength) < (uint)destination.Length)
+                {
+                    destination[computedOutputLength] = (byte)HexConverter.ToCharLower((int)value);
+                    value >>= 4;
+                }
+            }
+            else
             {
-                destination[computedOutputLength] = (byte)hexTable[(int)value & 0xf];
-                value >>= 4;
+                while ((uint)(--computedOutputLength) < (uint)destination.Length)
+                {
+                    destination[computedOutputLength] = (byte)HexConverter.ToCharUpper((int)value);
+                    value >>= 4;
+                }
```

Comment:
Several folks like @ahsonkhan spent a while tuning all of the code in this area.  It'd be good to validate these changes against our perf tests to ensure nothing regresses.

Reply:
The comment above says otherwise. Anyway, this seems to be for some reason pointed out more than once here so I changed the code to do at most 8 iterations instead of 16 which the code does today.

Reply:
> The comment above says otherwise.

That's not how I read it.  To me, the comment highlights that the author went to great lengths thinking about perf and the implications of how it was written, mentioning additional work that could be tried in the future.

Please run the perf tests to ensure it hasn't regressed.  If your changes make it faster, all the better.

Reply:
@stephentoub I misunderstood the comment and you are right the code was tuned. I tweaked the code and benchmarked the method `TryFormatUInt64X` with the following results on my machine (about 10% speedup for the longest casein this PR).

```
BenchmarkDotNet=v0.12.0, OS=macOS 10.15.2 (19C57) [Darwin 19.2.0]
Intel Core i7-7920HQ CPU 3.10GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha.1.20061.5
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.6301, CoreFX 4.700.19.56404), X64 RyuJIT
  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.6007, CoreFX 5.0.20.6007), X64 RyuJIT


|                    Method |                Input |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------------------- |--------------------- |----------:|----------:|----------:|------:|--------:|
| Existing_TryFormatUInt64X |                    1 |  9.682 ns | 0.2332 ns | 0.2685 ns |  1.00 |    0.00 |
|       PR_TryFormatUInt64X |                    1 |  9.219 ns | 0.2087 ns | 0.2143 ns |  0.95 |    0.04 |
|                           |                      |           |           |           |       |         |
| Existing_TryFormatUInt64X |                 1000 | 13.786 ns | 0.4389 ns | 1.2802 ns |  1.00 |    0.00 |
|       PR_TryFormatUInt64X |                 1000 | 11.812 ns | 0.2820 ns | 0.4306 ns |  0.93 |    0.07 |
|                           |                      |           |           |           |       |         |
| Existing_TryFormatUInt64X |               123456 | 14.584 ns | 0.8399 ns | 0.8625 ns |  1.00 |    0.00 |
|       PR_TryFormatUInt64X |               123456 | 14.661 ns | 0.3340 ns | 0.8379 ns |  1.05 |    0.11 |
|                           |                      |           |           |           |       |         |
| Existing_TryFormatUInt64X | 18446744073709551615 | 26.105 ns | 0.5646 ns | 1.5265 ns |  1.00 |    0.00 |
|       PR_TryFormatUInt64X | 18446744073709551615 | 24.166 ns | 0.5207 ns | 1.0518 ns |  0.91 |    0.06 |
```

@EgorBo I don't think ROS version will be ever faster. It has extra memory read which I guess will be hard to beat with the branch prediction in modern CPUs

Reply:
@marek-safar if https://github.com/dotnet/runtime/pull/1644 gets merged it will remove bound checks for the ROS version so it will be just `and` and a few movs:
```csharp
public static char ToCharUpper(int value) => 
	(char)(s_hexEncodeMap[value & 0xF]);
```
Codegen for ROS-version (in my branch):
```asm
; Method Program:ToCharUpper(int):ushort
       and      ecx, 15
       movsxd   rax, ecx
       mov      rdx, 0xD1FFAB1E
       movzx    rax, byte  ptr [rax+rdx]
       ret
```
vs codegen for this PR's impl:
```asm
       and      ecx, 15
       cmp      ecx, 9
       jg       SHORT G_M2907_IG04
       lea      eax, [rcx+48]
       jmp      SHORT G_M2907_IG05
G_M2907_IG04:
       lea      eax, [rcx+87]
G_M2907_IG05:
       movzx    rax, ax
       ret   
```
Not sure will it be faster though, I guess branch predictor + speculative execution indeed help here especially in benchmarks where input is usually constant.

Reply:
I measured the changes in this PR and it looks like there is some regression:
```C#
public class HexFormatTest
{
    [Params(-1, 0, 1, long.MaxValue, int.MaxValue, int.MinValue, long.MinValue)]
    public long Value;

    private byte[] _output;

    [Params('x', 'X')]
    public StandardFormat Format;

    [GlobalSetup]
    public void Setup()
    {
        _output = new byte[1_000];
    }

    [Benchmark(Baseline = true)]
    public bool Before()
    {
        return Utf8Formatter.TryFormat(Value, _output, out _, Format);
    }

    [Benchmark]
    public bool After()
    {
        return TryFormat(Value, _output, out _, Format);
    }
}
```
The only change/diff is:
```diff
- string hexTable = (useLower) ? FormattingHelpers.HexTableLower : FormattingHelpers.HexTableUpper;
- while ((uint)(--computedOutputLength) < (uint)destination.Length)
- {
-    destination[computedOutputLength] = (byte)hexTable[(int)value & 0xf];
-    value >>= 4;
- }

+ if (useLower)
+ {
+     while ((uint)(--computedOutputLength) < (uint)destination.Length)
+     {
+         destination[computedOutputLength] = (byte)ToCharLower((int)value);
+         value >>= 4;
+     }
+ }
+ else
+ {
+     while ((uint)(--computedOutputLength) < (uint)destination.Length)
+     {
+         destination[computedOutputLength] = (byte)ToCharUpper((int)value);
+         value >>= 4;
+     }
}
```

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha1-015914
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.19.56303, CoreFX 5.0.19.56306), X64 RyuJIT
  Job-AZDGGY : .NET Core 5.0.0 (CoreCLR 5.0.19.56303, CoreFX 5.0.19.56306), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  

```
| Method |                Value | Format |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |--------------------- |------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|------:|------:|------:|----------:|
| **Before** |          **-2147483648** |      **X** | **15.206 ns** | **0.3289 ns** | **0.3077 ns** | **15.181 ns** | **14.644 ns** | **15.717 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |          -2147483648 |      X | 20.875 ns | 0.4065 ns | 0.3802 ns | 20.901 ns | 20.090 ns | 21.486 ns |  1.37 |    0.04 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |          **-2147483648** |      **x** | **15.270 ns** | **0.2164 ns** | **0.1918 ns** | **15.310 ns** | **14.829 ns** | **15.493 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |          -2147483648 |      x | 22.338 ns | 0.4745 ns | 0.6495 ns | 22.388 ns | 21.488 ns | 24.055 ns |  1.46 |    0.05 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                   **-1** |      **X** | **15.203 ns** | **0.2947 ns** | **0.2756 ns** | **15.182 ns** | **14.801 ns** | **15.802 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                   -1 |      X | 17.541 ns | 0.2877 ns | 0.2691 ns | 17.637 ns | 16.929 ns | 17.795 ns |  1.15 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                   **-1** |      **x** | **15.204 ns** | **0.2914 ns** | **0.2726 ns** | **15.336 ns** | **14.726 ns** | **15.493 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                   -1 |      x | 18.770 ns | 0.3571 ns | 0.3340 ns | 18.887 ns | 18.174 ns | 19.304 ns |  1.23 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** | **-9223372036854775808** |      **X** | **15.364 ns** | **0.2791 ns** | **0.2474 ns** | **15.403 ns** | **14.830 ns** | **15.732 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After | -9223372036854775808 |      X | 17.920 ns | 0.3819 ns | 0.3573 ns | 18.005 ns | 17.098 ns | 18.494 ns |  1.17 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** | **-9223372036854775808** |      **x** | **16.198 ns** | **0.3513 ns** | **0.6060 ns** | **16.146 ns** | **15.195 ns** | **17.338 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After | -9223372036854775808 |      x | 16.324 ns | 0.3005 ns | 0.2952 ns | 16.374 ns | 15.836 ns | 16.821 ns |  1.03 |    0.05 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                    **0** |      **X** |  **5.907 ns** | **0.1366 ns** | **0.1277 ns** |  **5.933 ns** |  **5.685 ns** |  **6.183 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                    0 |      X |  5.321 ns | 0.1182 ns | 0.1106 ns |  5.365 ns |  5.133 ns |  5.529 ns |  0.90 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                    **0** |      **x** |  **5.802 ns** | **0.1224 ns** | **0.1085 ns** |  **5.834 ns** |  **5.632 ns** |  **6.004 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                    0 |      x |  6.393 ns | 0.1138 ns | 0.1009 ns |  6.403 ns |  6.151 ns |  6.573 ns |  1.10 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                    **1** |      **X** |  **5.870 ns** | **0.1346 ns** | **0.1259 ns** |  **5.901 ns** |  **5.620 ns** |  **6.078 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                    1 |      X |  5.371 ns | 0.1138 ns | 0.1064 ns |  5.386 ns |  5.118 ns |  5.518 ns |  0.92 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |                    **1** |      **x** |  **5.780 ns** | **0.0850 ns** | **0.0754 ns** |  **5.763 ns** |  **5.666 ns** |  **5.886 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |                    1 |      x |  6.223 ns | 0.1333 ns | 0.1246 ns |  6.276 ns |  6.048 ns |  6.368 ns |  1.08 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |           **2147483647** |      **X** | **10.822 ns** | **0.2222 ns** | **0.2078 ns** | **10.790 ns** | **10.510 ns** | **11.329 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |           2147483647 |      X | 13.118 ns | 0.1054 ns | 0.0880 ns | 13.127 ns | 12.926 ns | 13.276 ns |  1.21 |    0.02 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |           **2147483647** |      **x** | **11.118 ns** | **0.2521 ns** | **0.4610 ns** | **11.295 ns** | **10.237 ns** | **12.117 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |           2147483647 |      x | 12.773 ns | 0.2848 ns | 0.3048 ns | 12.855 ns | 12.345 ns | 13.197 ns |  1.18 |    0.06 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |  **9223372036854775807** |      **X** | **15.325 ns** | **0.2527 ns** | **0.2364 ns** | **15.346 ns** | **14.856 ns** | **15.694 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |  9223372036854775807 |      X | 21.039 ns | 0.3554 ns | 0.3324 ns | 21.013 ns | 20.432 ns | 21.605 ns |  1.37 |    0.03 |     - |     - |     - |         - |
|        |                      |        |           |           |           |           |           |           |       |         |       |       |       |           |
| **Before** |  **9223372036854775807** |      **x** | **15.123 ns** | **0.3338 ns** | **0.3428 ns** | **15.241 ns** | **14.651 ns** | **15.612 ns** |  **1.00** |    **0.00** |     **-** |     **-** |     **-** |         **-** |
|  After |  9223372036854775807 |      x | 21.597 ns | 0.4643 ns | 0.4968 ns | 21.527 ns | 20.895 ns | 22.589 ns |  1.43 |    0.05 |     - |     - |     - |         - |

Reply:
Here's the disassembly for before/after in case it is of interest:
https://www.diffchecker.com/Eimp7IRP
[sharplab.io link](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AQgK4BmpApgE4QDc6WuBAcsVNQJYDGdDOe+AErEAdsA7t8AYQD2UAA4cANtQDK1AG7cKvNAwDMuBNinYA3umxXsl6wHo72AILYIE+SuxKZMgNbF5bGAwACNPUhkqbC4ZEQ1qcREAc2wRNhDqCCCZbAALCgAPfFsrDjFqETAlaNiIYCYcAAlCgBVQlQAZGQB3amwAXmwAIkwkBH0AFgBWADYAdgAOAE5QrgATClIh+j00a2wy4AqqmpE6huxmgrawigBVeXk+wZGxydnFpadCKQARAFEAGLbEo2PbWeRUDgaMBHJgzbAhHzVFpUACegMiUFhdwAkmIZhMABoAfQA8ko1gAKYjeZLYGFKYgUAA0iLRcMhFC4HDcsTZSJk1WIEAoXV6VDZqnkYBEAB4QhyKAA+bAbOplWEcfnYGTEeqHdlHCAAdShwCOIgAlKCLOD9qUxNgwFxgMQqr8OEkOMBZKJ6oM/WIrp7vcAIFTGcyrTsHdZDTEFPqKGsyfr5PqOhRksBcgNsFSuTy+SIBoNVMERGswFQ1piqNjgPhmDIAApUbm87XW0FxqwAfmdrvdSlDPqDwF7fZA2AAsrDcvh5wVCx3i922S63R6veO9WIY+gp6VSAX1YktbF8Fmc3m5TUk0dU+nM9mkrmbfaHXa+w7FcazR9S180wWNf2sOA5mwUgqlFMC4wAXyPL99n/HRAItbN80TDMnzTYBcJvd9cngh08BwfJrnaCh8xpUVxWoK1sEHK4bk6Ho+hnVjqIeJ4qB2Y9sAcbAMLKFJcxovUCP1REXV8boazWA5S1zXlsGxEQRD6KpvG6LIJOwAApXEWmybAKCUDgNkE4SkVENYsi4fIuF8MTdRU/JdRfeoQjISgqHwbAWk84zTO6WIAHJ6gsqyaIMuyqwgGzHCc7lfHcoJPMooJqK8Hx/HkAUZN6aJZS8aEaO6H081zWFoMiVIemKXZf2E3F6jWGQdBEKLsAAR2IH0aOxXwaNFM44pyEU4s8mDzmymI4kybtoNEV1uyqH00WS+qolUrJ2FzGQlLcEQuBmuqqCkspKsiXwsliI0dE3Ks8moSqKAijtsHCqhXOSHa1g4NdgCUNEfuq1IOBCW4IEC4L3vUsBwYyXVHkiN0RC23aDgUK74nYJ1FVcJ4eSqABaHauEgNypNwrw31zB7TwQNkJjZBYXqUpAEVlJSqFELH6QkqA1R6UsZHkcQoA4AAvFMqeOmiZSZ5rBO6XJlBoqkaUOK0qXJ8mcOTZ9pOAIiP2we9dYPc9NXEK8LdyT9fx/cCrDtyoHZEABtY28O8p2AF1aLQq1KLYigfapPWoxogAybBMAKUgg9I384+wZVlUGCZ0+sJCUIgqDgAFihSML/ZQR92cKCOtZcQUJQqVr+vG48Mkpe7OGnCSJIOwgNx4nxSyhaSK0g9BOBDDwBFDQnENd3DGk6RSOOXesN240ggtCWwcnsEIH1O+oS8zmvCgwCB5IAC1qBkCdIyqZlsBgbAkCYgBqbB9CY7PsAQBXUEoJITQlhDRWeiJkRBXRPWRseICTEhXrENez9WRPWwEWLsOpBTCnohxSU2BpSygVEqVUnsz5sikspHySpTTmktBvKwW8HSGi3COMcvp9wBhMNwxeYYIzr3zo6eo/sUz4UIozPMgxVydhLGWIhlZqy1jgbCZsbY1zYJ7EXOMg52E7jDBOQS+wZzzlzEuMAK4sElk3MOAxe5/SHhan2Dgp4qQUO9hfW8VsHy4XEYHKRTDvzGOsGhehQEsKDFAiEqwO8YJKDgoJQugkwkYWAoMMRptJG3gEs4uMwlRLC08nTGSIQ5IKVrMpTKakNJaSiDpHo+kQomTMjFayOihKOASg5aIzl/opEegZEpPk/LUHhs0sKkVoqWQ2JlGi3SkodOEqlFyGUDLZWCLcPKfgAhFXqCVampZLLxAhrmTKdUIhRBEE1ZCrVHDtTFt1XqA0hpI1Gq4bMoozLTTmdBSA9QFqxHiDQFapA1re02sAbaSzHCXOqQdOuuRjquDKOdc59Qrr6huj9O6D1SxhK5m9b6vQvq3T+mJQGwNuSg3BlVM5WMYYqDhkFfI31sQo0kujKgmNsZwvcPjCghNaEk07BTKmNN6TDIZreZmAC2Yc0JTzZ0r0BaaTciLMW3QJZdxlvLNYitZkq1yHDW5Li3HTQYlQIJ+wWG/g1lrAsNtgD60NpkiRr5bxMWtoNW2OgLyeKdtavstr3Zqj9fbbsfs5B+KyR64iIcZFhxaA/XINZLU61jmgpxoaHSZ2zrnYRDpK59mLXGCyopBIhrjPazwOsfXOoNkbaNJt3Xm0CT4p1VoPHdi8cRINcYq3gW7bEKNj5/Fm2DqHJUVpk1SFTVQXi1AM0HiETE3NaCs452wHnNdpai2mu3iXMuQCvzV1bkihuTcW6IuOu3JQndvY9z7gPIeFAR5lDEhPKeM8lVORrEFFNNZF1UBjk6detpBKZ3jlEgogJC07ypH+q1T8mQ0VVEsZiDIN0HyQDgb+EUnARWwDOTO+HMARWzQXA92Aa43svR4a9bcm4Pu7vgXu/cdCvvfWPL9X5p7wl6f+2d8702GnA1+QdUGYNwcEghpD+s83YAw4OTOOG8PYAimAIjJGN1kYoyehCQA==)

Reply:
cc @tannergooding

Reply:
Is there now minimal perf difference for this API based on this PR or are we still seeing noticeable regressions?